src/env.cc

```diff
@@ -1857,63 +1842,60 @@ size_t Environment::NearHeapLimitCallback(void* data,
     if (strcmp(stats.space_name(), "new_space") == 0 ||
         strcmp(stats.space_name(), "new_large_object_space") == 0) {
       young_gen_size += stats.space_used_size();
-    } else {
+      // Ignore code spaces for the calculation.
+    } else if (strstr(stats.space_name(), "code") == nullptr) {
       old_gen_size += stats.space_used_size();
     }
-  }
-
-  Debug(env,
-        DebugCategory::DIAGNOSTICS,
-        "max_young_gen_size=%" PRIu64 ", "
-        "young_gen_size=%" PRIu64 ", "
-        "old_gen_size=%" PRIu64 ", "
-        "total_size=%" PRIu64 "\n",
-        static_cast<uint64_t>(max_young_gen_size),
-        static_cast<uint64_t>(young_gen_size),
-        static_cast<uint64_t>(old_gen_size),
-        static_cast<uint64_t>(young_gen_size + old_gen_size));
-
-  uint64_t available = GuessMemoryAvailableToTheProcess();
-  // TODO(joyeecheung): get a better estimate about the native memory
-  // usage into the overhead, e.g. based on the count of objects.
-  uint64_t estimated_overhead = max_young_gen_size;
-  Debug(env,
-        DebugCategory::DIAGNOSTICS,
-        "Estimated available memory=%" PRIu64 ", "
-        "estimated overhead=%" PRIu64 "\n",
-        static_cast<uint64_t>(available),
-        static_cast<uint64_t>(estimated_overhead));
-
-  // This might be hit when the snapshot is being taken in another
-  // NearHeapLimitCallback invocation.
-  // When taking the snapshot, objects in the young generation may be
-  // promoted to the old generation, result in increased heap usage,
-  // but it should be no more than the young generation size.
-  // Ideally, this should be as small as possible - the heap limit
+    total_size += stats.space_used_size();
+  }
+
+  // There are at least two things that can lead to memory overhead
+  // during the heap snapshot generation.
+  // 1. Heap snapshot generation triggers garbage collection for accuracy.
+  //    When objects get promoted from the young generation to the old
+  //    generation, memory usage in the old space can be increased.
+  // 2. V8 calculates and caches the line ends for scripts found in the heap
+  //    in order to produce line/column numbers for the source locations.
+  //    The allocation of these caches also increases the memory usage.
+  // The estimate here is a guess that combines the two factors we've noticed
+  // so far. Ideally, this should be as small as possible - the heap limit
   // can only be restored when the heap usage falls down below the
   // new limit, so in a heap with unbounded growth the isolate
   // may eventually crash with this new limit - effectively raising
   // the heap limit to the new one.
-  size_t new_limit = current_heap_limit + max_young_gen_size;
+  size_t estimated_overhead = young_gen_size + (old_gen_size / 2);
+  size_t new_limit = current_heap_limit + estimated_overhead;
```

Comment:
Not sure about this formula - maybe we should use inital_heap_limit here as the base instead of avoid unbound growth of the new limit